### ISSUE TITLE
Ranges and Lists

### ISSUE THREAD

**Issue by dazinator** [kthompson/glob#12]
Currently range and lists don't appear to be supported. i.e these tests will fail:

```     

        [Fact]
        public void CanMatchSingleFileUsingCharRange()
        {
            var glob = new Glob("*fil[e-z].txt");
            Assert.True(glob.IsMatch("bigfile.txt"));
            Assert.True(glob.IsMatch("smallfilf.txt"));
            Assert.False(glob.IsMatch("smallfila.txt"));
            Assert.False(glob.IsMatch("smallfilez.txt"));
        }

        [Fact]
        public void CanMatchSingleFileUsingNumberRange()
        {
            var glob = new Glob("*file[1-9].txt");
            Assert.True(glob.IsMatch("bigfile1.txt"));
            Assert.True(glob.IsMatch("smallfile8.txt"));
            Assert.False(glob.IsMatch("smallfile0.txt"));
            Assert.False(glob.IsMatch("smallfilea.txt"));
        }

       [Fact]
        public void CanMatchSingleFileUsingCharList()
        {
            var glob = new Glob("*file[abc].txt");
            Assert.True(glob.IsMatch("bigfilea.txt"));
            Assert.True(glob.IsMatch("smallfileb.txt"));
            Assert.False(glob.IsMatch("smallfiled.txt"));
            Assert.False(glob.IsMatch("smallfileaa.txt"));
        }
```

I also noticed that you rely on Regex and this made me wonder whether an implementation that didn't use Regex would be better.. To that end I have written a tokeniser that just uses a string reader. I am now working on Regexless matcher that will evaluate the tokenised glob, against an input string. However once I have finished this I would have pretty much re-imple,ented the guts of this library - but I am just too curious to know how it will stack up against a Regex implementation!

**Comment by kthompson** [comment]
There is already a tokenizer to get the individual tokens. You could write
an evaluator that did not convert the parse tree to regex instead. I
imagine regex would be fastest but I have no definitive evidence to show
that.

**Comment by dazinator** [comment]
Ok so I ended up writing my own regexless globbing library (sorry!)..

I just thought it would be an interesting excercise, and I also wanted to see what the performance differences would be like - would regex be faster?

Parsing this glob:

```
var glob= new global::Glob.Glob("p?th/*a[bcd]b[e-g]a[1-4][!wxyz][!a-c][!1-3].*");
```

took the current regex version: "00:00:00.0000060"
took my regexless version: "00:00:00.0005302"

So I am massively slower there. Will have to figure out whats going on...

However, once the glob was parsed, doing 1000 matches against it:

``` 
var result = comparisonGlob.IsMatch("pAth/fooooacbfa2vd4.txt");
```

took the current regex version: 00:00:00.0669022
took the regexless version: 00:00:00.0276693

So I kind of won in that department but it was short lived as I noticed this option:

```
 comparisonGlob = new global::Glob.Glob("p?th/*a[bcd]b[e-g]a[1-4][!wxyz][!a-c][!1-3].*", GlobOptions.Compiled);

```

Compiling the regex made the parsing take quite a bit slower than the regexless version

regex: 00:00:00.0304611
regexless: 00:00:00.0005302

And then 1000 matches:

```
		regex:	"00:00:00.0422586"
                regexless:	"00:00:00.0276693"

```

So... to be honest, these are very small numbers. But I am going to take that as a small victory :-)

**Comment by dazinator** [comment]
Ahh shortlived..

Testing today with a 1000 randomly generated string matches, and regex is kicking ass..

Regex seems better than I thought.. :-)

**Comment by kthompson** [comment]
Haha. Not sure if you saw but I did a couple things to help performance in
the parser. Should be in Nuget already.

**Comment by dazinator** [comment]
Ok so I revamped my code, tweaked some things, ticked the optimise box, did a rain dance etc, prayed to the cpu gods,

Downloaded your latest release, and did some comparison tests. Basically in the test, I create a glob, give it a few warmup passes, and I generate 10000 random strings that match that glob pattern. I then time how long it takes for the glob to successfully match those 10000 strings with a stopwatch. I have repeated this test for a few different globb patterns that vary in complexity.

The outcome, for the majority of the time, is that I can now beat regex, but the margins of victory are so ridiculously narrow, as to be mostly meaningless.. still it gives me the warm fuzzies.

``` regexless wins by:  00:00:00.0007434, matching 10000 against p?th/a[bcd]b[e-g]a[1-4][!wxyz][!a-c][!1-3].txt  ```
``` regexless wins by:  00:00:00.0005759 matching 10000 against p?th/a[bcd]b[e-g]a[1-4][!wxyz][!a-c][!1-3].txt  ```

So the outcome basically is.. regex is "good enough", but I didn't completely waste my effort because I achieved my goal :-)

If you want to compare yourself (I saw you were using benchmarkdotnet which looks pretty good) I have released my version (it's a pre-release nuget package at the mo) here: http://www.nuget.org/packages/DotNet.Glob/1.2.0-unstable0001
